<a href="https://colab.research.google.com/github/gimoralesg/AlgorithmsUN2021II/blob/main/Lab7/grupo28_AlgorithmicToolboxWeek1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [ ]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Writing APlusB.cpp


In [ ]:
!g++ APlusB.cpp -o APlusB
!./APlusB

4 10
14

### Test input output files

In [ ]:
%%writefile input1.txt
4 8

Writing input1.txt


In [ ]:
%%writefile output1.txt
12

Overwriting output1.txt


In [ ]:
%%writefile input2.txt
-5 9

Overwriting input2.txt


In [ ]:
%%writefile output2.txt
4

Overwriting output2.txt


In [ ]:
!./APlusB < input1.txt

12

In [ ]:
!./APlusB < input2.txt

4

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [ ]:
!grep -xvFf output1.txt output1.txt 

In [ ]:
!grep -xvFf output1.txt output2.txt 

4


In [ ]:
!grep -xvFf output2.txt output1.txt 

12


In [ ]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [ ]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Overwriting sum.py


In [ ]:
!python sum.py

32 90
122


In [ ]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [ ]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

In [ ]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Overwriting max_pairwise_product.cpp


In [ ]:
!g++ max_pairwise_product.cpp -o max_pairwise_product
!./max_pairwise_product

6
23 1 89 56 34 5
4984


### Test 1: Small numbers

In [ ]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Writing input1.txt


In [ ]:
!./max_pairwise_product < input1.txt

90


### Test 2: largest input numbers 200000 200000 in the input

In [ ]:
%%writefile input2.txt
6
1 3 200000 9 200000 23

Writing input2.txt


In [ ]:
!./max_pairwise_product < input2.txt

1345294336


Changing the variables to long long

In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
			       aux_product =  (long long) numbers[first] * numbers[second];
            if ( max_product < aux_product)  {
				        max_product = aux_product;
             }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Overwriting max_pairwise_product_long.cpp


In [ ]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!./max_pairwise_product_long < input1.txt
!./max_pairwise_product_long < input2.txt

90
40000000000


### Test 3  stress tests 20  times  random n in [2, 20] 

In [ ]:
%%writefile stress_test_3.cpp
#include <iostream>
#include <vector>
#include <cstdlib>

using std::cin;
using std::cout;
using std::vector;

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
			       aux_product =  (long long) numbers[first] * numbers[second];
            if ( max_product < aux_product)  {
				        max_product = aux_product;
             }
        }
    }

    return max_product;
}

long long MaxPairwiseProductFast(const std::vector<int>& numbers){
	int n = numbers.size();
	
	int max_index1 = -1;
	for(int i=0; i<n; i++){
		if((max_index1 == -1) || (numbers[i] > numbers[max_index1]))
			max_index1 = i;
	}
	
	int max_index2 = -1; 
	for(int j=0; j<n; j++){
		if((j != max_index1) && ((max_index2 == -1) || (numbers[j] > numbers[max_index2])))
			max_index2 = j;
	}
	
	cout << max_index1 << " "<< max_index2 << "\n";
	
	return((long long)(numbers[max_index1])) * numbers[max_index2];
}

int main() {
    
    int j = 0;
    while(j<20){
      int n = rand() % 19 + 2;
      cout << n << "\n";
      vector<int> a;
      
      for(int i =0; i<n; ++i){
        a.push_back(rand() % 100000);
      }
      for(int i =0; i<n;++i){
        cout << a[i] << ' ';
      }
      
      std::cout << "\n";
      
      long long res1 = MaxPairwiseProduct(a);
      long long res2 = MaxPairwiseProductFast(a);
      
      if(res1 != res2){
        cout << "Wrong answer: " << res1 << " " << res2 << "\n";
        break;
      }else{
        cout << "OK\n"; 
      }
      j++;
    }
    
    int n; 
    cin >> n;
    vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        cin >> numbers[i];
    }

    //std::cout << MaxPairwiseProduct(numbers) << "\n";
    long long result = MaxPairwiseProductFast(numbers);
	  cout << result << "\n";
	return 0;
}

Writing stress_test_3.cpp


In [ ]:
!g++ stress_test_3.cpp -o stresstest3

In [ ]:
!./stresstest3

4
30886 92777 36915 47793 
1 3
OK
10
85386 60492 16649 41421 2362 90027 68690 20059 97763 13926 
8 5
OK
9
83426 89172 55736 5211 95368 2567 56429 65782 21530 
4 1
OK
8
65123 74067 3135 13929 79802 34022 23058 33069 
4 1
OK
7
61393 18456 75011 78042 76229 77373 84421 
6 3
OK
3
13784 98537 75198 
1 2
OK
11
98315 64370 66413 3526 76091 68980 59956 41873 6862 99170 6996 
9 0
OK
18
2305 20925 77084 36327 60336 26505 50846 21729 61313 25857 16124 53895 19582 545 98814 33367 15434 90364 
14 17
OK
2
13750 71087 
1 0
OK
15
17276 47178 95788 93584 5403 2651 92754 12399 99932 95060 49676 93368 47739 10012 36226 
8 2
OK
7
48094 97539 40795 80570 51434 60378 97467 
1 6
OK
15
10097 12902 73317 70492 26652 60756 97301 60280 24286 9441 53865 29689 28444 46619 58440 
6 2
OK
13
58031 8117 38097 5771 34481 90675 20709 98927 4567 77856 79497 72353 54586 
7 5
OK
6
55306 64683 6219 28624 51528 32871 
1 0
OK
17
48829 9503 30019 58270 63368 59708 86715 26340 18149 47796 723 42618 2245 22846 93451 92921 43555 

### Test 4  time test  measure the execution time running with the the largest worst case file (200000 increasing numbers, 200000 and 200000 the end) 

In [ ]:
%%writefile Test2Write.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input3.txt");

 MyFile << 200000 <<"\n";  // n 
 for (int i=1; i < 200001; i++){
     MyFile << i <<" ";
 }
  
  // Close the file
  MyFile.close();
}

Overwriting Test2Write.cpp


In [ ]:
!g++ Test2Write.cpp -o Test2Write
!./Test2Write

Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [ ]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long < input3.txt

39999800000

real	1m57.837s
user	1m57.568s
sys	0m0.039s


##### Calculate  the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line 

```cpp
 int main() {                         //N inst exec by line executed  
   
    int n;                            //1                      
    std::cin >> n;                    //1
                                              
    std::vector<int> numbers(n);      //n    init vector
    .
    .
 }

```

In [ ]:
%%writefile test_init_vectors.cpp
#include <iostream>
#include <vector>
#include <algorithm>
int main() {                                                               
    int n = 10;                                                                                                                            //1
    std::vector<int> numbers(n);                                            
    for (int i = 0; i < n; ++i) {                                           
        std::cout << numbers[i];                                             
    }

    std::cout << "\n";                       
    return 0;                                                              
}  

Overwriting test_init_vectors.cpp


In [ ]:
!g++ test_init_vectors.cpp -o test_init_vectors
!./test_init_vectors

0000000000


In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //N inst exec by line executed 
    long long max_product = 0LL;                                             //1                                          
	  long long aux_product = 0LL;                                             //1 
    int n = numbers.size();                                                  //1 
                                                                             
    for (int first = 0; first < n; ++first) {                                //1+2n  
        for (int second = first + 1; second < n; ++second) {                 //n + (n-1)n 
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2 
            if (max_product < aux_product)  {                                //(n-1)n/2 
				        max_product = aux_product;                                   //[0, (n-1)n/2] 
             }
        }
    }
                                                                             // -------------- 
    return max_product;                                                      // 4 + 3n + [2(n-1)n, 2.5 (n-1)n] 
}

int main() {                                              //N inst exec by line executed 
    int n;                                                //1  
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]
    return 0;                                             // -------------- 
}                                                         // 9 + 7n + [2(n-1)n, 2.5 (n-1)n]


Overwriting max_pairwise_product_long.cpp


Count executed intructions in code and verify with inputs 1, 2 and 3

In [ ]:
%%writefile max_pairwise_product_long_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {              //N inst exec by line executed 
    
    long long m = 0LL;                                                       //1
    
    long long max_product = 0LL;                                             //1
    execints++;

	  long long aux_product = 0LL;                                             //1
    execints++;

    int n = numbers.size();                                                  //1 
    execints++;

    execints++;                                                                         
    for (int first = 0; first < n; ++first) {                                //1+2n  
        execints += 2;

        long long m = 0LL;

        execints++;                                    
        for (int second = first + 1; second < n; ++second) {                 //n+ (n-1)n 
            execints += 2;                                                

			      aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2 
            execints++; 

            execints++; 
            if ( max_product < aux_product)  {                               //(n-1)n/2 
                       
				        max_product = aux_product;                                   //[0, (n-1)n/2] 
                execints++; 
                m++;
             }
        }
        if ( (first <= 10) || (n- first <= 10)) std::cout << m << "\n";
    }
                                                                            // -------------- 
    return max_product;                                                     // 4 + 3n + [2(n-1)n, 2.5 (n-1)n] 
}

int main() {                                                 //N inst exec by line executed 
    long long execints =0LL;

    int n;                                                   //1
    execints++;                                                              
    
    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;
    
    execints++;
    for (int i = 0; i < n; ++i) {                            //1 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n
    }
    
    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";                       
    return 0;                                                // -------------- 
}                                                            // 9 + 7n + [2(n-1)n, 2.5 (n-1)n]

Overwriting max_pairwise_product_long_execints.cpp


In [ ]:
!g++ max_pairwise_product_long_execints.cpp -o max_pairwise_product_long_execints
!./max_pairwise_product_long_execints < input1.txt
!./max_pairwise_product_long_execints < input2.txt
!./max_pairwise_product_long_execints < input3.txt

9
5
4
3
2
2
2
2
1
0
289
2
1
2
0
0
0
116
199999
100000
66667
50000
40000
33334
28572
25000
22223
20000
18182
2
2
2
2
2
2
2
2
1
0
80003672077


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T = 9 + 7n + [2(n-1)n, 2.5 (n-1)n]$$

It depends on the size of the lists and  the order of the list.

For the input1.txt 
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 9 + 7(10) + 2(9)(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 289 $$

For the input2.txt 
```
6
1 3 200000 9 200000 23
```
$$T = 9 + 7(6) + 2(5)(6)  +  2 + 1 + 2 + 0 + 0 + 0 = 116 $$


For the input3.txt that is the worst case (n=200000 and list in incresing order)

```
200000
1 2 3 4 5 6 7 8 9 10 ... 199998 199999 200000
```

$$
\begin{array}{ll}
T &= 9 + 7(200000) + (2)(199999)(200000)  + (199999 +
100000 + 66667 +50000 \cdots 2 + 2 + 1)\\
&= 80003672077  \\
  &< 9 + 7(200000) + (2.5)(199999)(200000) \\
  & = 100000900009
\end{array}
  $$

Calculate estimated time by instruction and number of instructions by second

In [ ]:
# Estimated time by instruction

128/80003672077 


1.5999265618308827e-09

In [ ]:
# Estimated number of instructions by second

1/(128/80003672077) 

625028688.1015625

### Efficient solution

Add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile effsol_maxpairwise.cpp
#include <iostream>
#include <vector>
#include <cstdlib>

using std::cin;
using std::cout;
using std::vector;


long long MaxPairwiseProductFast(const std::vector<int>& numbers){
	int n = numbers.size();                                                                           //1
	
	int max_index1 = -1;                                                                              //1
	for(int i=0; i<n; i++){                                                                           //1+2n
      if((max_index1 == -1) || (numbers[i] > numbers[max_index1]))                                  //n
          max_index1 = i;
	}
	
	int max_index2 = -1;                                                                              //1
	for(int j=0; j<n; j++){                                                                           //1+2n
      if((j != max_index1) && ((max_index2 == -1) || (numbers[j] > numbers[max_index2])))           //n
          max_index2 = j;
	}
	
	return((long long)(numbers[max_index1])) * numbers[max_index2];                                   //6+4n[n,2n]
}

int main() {
    
    int n;                                                             //1
    cin >> n;                                                          //1
    vector<int> numbers(n);                                            //n
    for (int i = 0; i < n; ++i) {                                      //1+2n
        cin >> numbers[i];                                             //n
    }
    long long result = MaxPairwiseProductFast(numbers);                //1+6+4n[n,2n]
	  cout << result << "\n";                                            //1
	return 0;
}

Writing effsol_maxpairwise.cpp


In [ ]:
!g++ effsol_maxpairwise.cpp -o effsol_maxpairwise

Test with inputs 1 and 2



In [ ]:
!./effsol_maxpairwise < input1.txt
!./effsol_maxpairwise < input2.txt

9 8
90
2 4
40000000000


Test and time it with input 3

In [ ]:
!time ./effsol_maxpairwise < input3.txt

199999 199998
39999800000

real	0m0.054s
user	0m0.052s
sys	0m0.002s


20 stress tests

In [ ]:
%%writefile effsol_maxpairwise.cpp
#include <iostream>
#include <vector>
#include <cstdlib>

using std::cin;
using std::cout;
using std::vector;


long long MaxPairwiseProductFast(const std::vector<int>& numbers){
	int n = numbers.size();                                                                           //1
	
	int max_index1 = -1;                                                                              //1
	for(int i=0; i<n; i++){                                                                           //1+2n
      if((max_index1 == -1) || (numbers[i] > numbers[max_index1]))                                  //n
          max_index1 = i;
	}
	
	int max_index2 = -1;                                                                              //1
	for(int j=0; j<n; j++){                                                                           //1+2n
      if((j != max_index1) && ((max_index2 == -1) || (numbers[j] > numbers[max_index2])))           //n
          max_index2 = j;
	}
	
	return((long long)(numbers[max_index1])) * numbers[max_index2];                                   //6+4n[n,2n]
}

int main() {
    /*
    int j = 0;
    while(j<20){
      int n = rand() % 19 + 2;
      cout << n << "\n";
      vector<int> a;
      
      for(int i =0; i<n; ++i){
        a.push_back(rand() % 100000);
      }
      for(int i =0; i<n;++i){
        cout << a[i] << ' ';
      }
      
      std::cout << "\n";
      
      long long res2 = MaxPairwiseProductFast(a);
      
      cout << res2 << "\n";
      j++;
    }*/
    
    int n;                                                             //1
    cin >> n;                                                          //1
    vector<int> numbers(n);                                            //n
    for (int i = 0; i < n; ++i) {                                      //1+2n
        cin >> numbers[i];                                             //n
    }
    long long result = MaxPairwiseProductFast(numbers);                //1+6+4n[n,2n]
	  cout << result << "\n";                                            //1
	return 0;
}

Overwriting effsol_maxpairwise.cpp


In [ ]:
!g++ effsol_maxpairwise.cpp -o effsol_maxpairwise

In [ ]:
!./effsol_maxpairwise

4
30886 92777 36915 47793 
4434091161
10
85386 60492 16649 41421 2362 90027 68690 20059 97763 13926 
8801309601
9
83426 89172 55736 5211 95368 2567 56429 65782 21530 
8504155296
8
65123 74067 3135 13929 79802 34022 23058 33069 
5910694734
7
61393 18456 75011 78042 76229 77373 84421 
6588383682
3
13784 98537 75198 
7409785326
11
98315 64370 66413 3526 76091 68980 59956 41873 6862 99170 6996 
9749898550
18
2305 20925 77084 36327 60336 26505 50846 21729 61313 25857 16124 53895 19582 545 98814 33367 15434 90364 
8929228296
2
13750 71087 
977446250
15
17276 47178 95788 93584 5403 2651 92754 12399 99932 95060 49676 93368 47739 10012 36226 
9572286416
7
48094 97539 40795 80570 51434 60378 97467 
9506833713
15
10097 12902 73317 70492 26652 60756 97301 60280 24286 9441 53865 29689 28444 46619 58440 
7133817417
13
58031 8117 38097 5771 34481 90675 20709 98927 4567 77856 79497 72353 54586 
8970205725
6
55306 64683 6219 28624 51528 32871 
3577357998
17
48829 9503 30019 58270 63368 59708 86715 2634

Count executed instructions in code and verify with inputs 1,2 and 3


In [ ]:
!g++ effsol_maxpairwise.cpp -o effsol_maxpairwise
!./effsol_maxpairwise < input1.txt
!./effsol_maxpairwise < input2.txt
!./effsol_maxpairwise < input3.txt

90
40000000000
39999800000


Calculate estimated time by instruction and number of instructions by second

### Super efficient solution

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line 

In [ ]:
%%writefile supereffsol.cpp
#include <iostream>

using namespace std;

int main() {
    long long n; 
    long long auxmax;
    long long max1 = 0; 
    long long max2 = 0;

    cin >> n;
    
    for (int i=0; i < n; i++){                                       
        cin >> auxmax;
        
        if (auxmax >= max1){                                        
            max2 = max1;                                     
            max1 = auxmax;
        }else if(max2 == 0 || auxmax >= max2){                  
            max2 = auxmax;                                  
        }
    }
    cout << max1 * max2<<"\n";
    
    return 0;
}

Overwriting supereffsol.cpp


In [ ]:
!g++ supereffsol.cpp -o supereffsol
!./supereffsol

5 
1 2 3 4 5 
20


Test with inputs 1 and 2


 

In [ ]:
!g++ supereffsol.cpp -o supereffsol
!./supereffsol < input1.txt
!./supereffsol < input2.txt

90
40000000000


Test and time it with input 3

In [ ]:
!time ./supereffsol < input3.txt

39999800000

real	0m0.055s
user	0m0.052s
sys	0m0.001s


20 stress tests 

In [ ]:
%%writefile supereffsol_stress.cpp
#include <iostream>

using namespace std;

int main() {
    long long n; 
    long long auxmax;
    long long max1 = 0; 
    long long max2 = 0;

    int j = 0;
    while(j<20){
        int n = rand() % 19 + 2;  
        cout << n << "\n";   
        
        for (int i=0; i < n; i++){                              
          auxmax = (rand() % 100000);
          cout << auxmax << " ";
          
          if (auxmax >= max1){                                        
              max2 = max1;                                     
              max1 = auxmax;
          }else if(max2 == 0 || auxmax >= max2){                  
              max2 = auxmax;                                  
          }
      }
      j++;
      cout << "\n"<< max1 * max2<<"\n";
    } 
    return 0;
}

Writing supereffsol_stress.cpp


In [ ]:
!g++ supereffsol_stress.cpp -o supereffsol_stress
!./supereffsol_stress

4
30886 92777 36915 47793 
4434091161
10
85386 60492 16649 41421 2362 90027 68690 20059 97763 13926 
9070157851
9
83426 89172 55736 5211 95368 2567 56429 65782 21530 
9323461784
8
65123 74067 3135 13929 79802 34022 23058 33069 
9323461784
7
61393 18456 75011 78042 76229 77373 84421 
9323461784
3
13784 98537 75198 
9633272731
11
98315 64370 66413 3526 76091 68980 59956 41873 6862 99170 6996 
9771914290
18
2305 20925 77084 36327 60336 26505 50846 21729 61313 25857 16124 53895 19582 545 98814 33367 15434 90364 
9799384380
2
13750 71087 
9799384380
15
17276 47178 95788 93584 5403 2651 92754 12399 99932 95060 49676 93368 47739 10012 36226 
9910256440
7
48094 97539 40795 80570 51434 60378 97467 
9910256440
15
10097 12902 73317 70492 26652 60756 97301 60280 24286 9441 53865 29689 28444 46619 58440 
9910256440
13
58031 8117 38097 5771 34481 90675 20709 98927 4567 77856 79497 72353 54586 
9910256440
6
55306 64683 6219 28624 51528 32871 
9910256440
17
48829 9503 30019 58270 63368 59708 86715 263

Count executed instructions in code and verify with inputs 1,2 and 3

In [ ]:
%%writefile supereffsol_execints.cpp
#include <iostream>

using namespace std;

int main() {
    long long execints = 0LL;

    long long n; 
    execints++;
    long long auxmax;
    execints++;
    long long max1 = 0; 
    execints++;
    long long max2 = 0;
    execints++;

    cin >> n;
    execints++;
    
    for (int i=0; i < n; i++){  
        execints += 2;                                     
        cin >> auxmax;
        execints++;
        
        if (auxmax >= max1){                                        
            max2 = max1;                                     
            max1 = auxmax;
            execints++;
        }else if(max2 == 0 || auxmax >= max2){                  
            max2 = auxmax;      
            execints++;                            
        }
    }
    cout << max1 * max2<<"\n";
    execints++;

    cout <<"executed instructions: "<< execints << "\n";
    
    return 0;
}

Overwriting supereffsol_execints.cpp


In [ ]:
!g++ supereffsol_execints.cpp -o supereffsol_execints
!./supereffsol_execints

5
2 4 7 5 80
560
executed instructions: 26


In [ ]:
!g++ supereffsol_execints.cpp -o supereffsol_execints
!./supereffsol_execints < input1.txt
!./supereffsol_execints < input2.txt
!./supereffsol_execints < input3.txt

90
executed instructions: 46
40000000000
executed instructions: 29
39999800000
executed instructions: 800006


Calculate estimated time by instruction and number of instructions by second

## Python

Efficient cython file reading, string parsing, and array building

https://stackoverflow.com/questions/37219875/efficient-cython-file-reading-string-parsing-and-array-building



File PreProcessing With C and Cython

https://github.com/EveryTimeIWill18/Cython_Repo/blob/master/FastFileProcessingWithCython.ipynb



In [ ]:
%%writefile effsol_python.py
def max_pairwise_product_Fast(numbers):
    n = len(numbers)  
    max_index1 = 0

    for i in range(n):
      if((max_index1 == 0) or (numbers[i] > numbers[max_index1])):
        max_index1 = i


    max_index2 = 0
    for j in range (n):
      if((j != max_index1) and ((max_index1 == 0) or (numbers[j] > numbers[max_index2]))):
        max_index2 = j

    return(numbers[max_index1] * numbers[max_index2])

if __name__ == '__main__':
  input_n = int(input())
  input_numbers = [int(x) for x in input().split()]
  print(max_pairwise_product_Fast(input_numbers))

Writing effsol_python.py


Test with inputs 1, 2 and 3

In [ ]:
!python effsol_python.py < input1.txt
!python effsol_python.py < input2.txt

90
40000000000


In [ ]:
!time python effsol_python.py < input3.txt

39999800000

real	0m0.181s
user	0m0.142s
sys	0m0.038s


20 stress test

In [ ]:
import random

def max_pairwise_product(numbers):
    n = len(numbers)
    max_product = 0
    for first in range(n):
        for second in range(first + 1, n):
            max_product = max(max_product,
                numbers[first] * numbers[second])
    return max_product

def max_pairwise_product_Fast(numbers):
    n = len(numbers)  
    max_index1 = 0

    for i in range(n):
      if((max_index1 == 0) or (numbers[i] > numbers[max_index1])):
        max_index1 = i


    max_index2 = 0
    for j in range (n):
      if((j != max_index1) and ((max_index1 == 0) or (numbers[j] > numbers[max_index2]))):
        max_index2 = j 

    #print (max_index1, " " , max_index2, "\n" )
    return(numbers[max_index1] * numbers[max_index2])


if __name__ == '__main__':

  j = 0
  while(j<20):
    n = (random.randint(2,20))
    print(n)
    a = []

    for i in range(n):
      a.append(random.randint(0,100000))
    
    for j in range(n):
      print(a[j], end =" ")
    
    res1 = max_pairwise_product(a)
    res2 = max_pairwise_product_Fast(a)

    if(res1 != res2):
      print("Wrong answer: ", res1," ", res2, "\n")
    else:
      print("\nOK\n")
    
    j+=1
    

  #input_n = int(input())
  #input_numbers = [int(x) for x in input().split()]
  #print(max_pairwise_product_Fast(input_numbers))

2
32760 45180 
OK

6
64107 80845 99791 24138 45660 3046 
OK

12
90431 73716 46424 51280 29362 88781 97069 22000 38274 88665 60696 63613 
OK

8
725 94875 19431 11098 27162 15867 58033 42719 
OK

16
39568 63686 63046 87296 92799 53566 9814 65595 35292 65638 26632 2489 64446 62807 48703 27609 
OK

9
22142 9893 83044 4041 20201 24549 73995 44331 99248 
OK

4
95640 2256 25406 25965 
OK

15
10661 71815 36158 69384 92405 87898 37894 78791 60628 84157 21946 22065 68502 8304 67139 
OK

15
59865 8059 71126 41968 96901 33367 45582 31215 51923 4444 72495 29900 18969 4589 16761 
OK

16
21587 28742 11947 49741 25891 4273 1005 37931 33290 51544 57105 63022 69949 29386 5413 19249 
OK

11
56603 42715 64117 24194 76894 73000 9607 84807 90730 11123 75414 
OK

8
78378 2785 42640 49850 53837 25533 49456 52135 
OK

16
3713 94689 29580 11059 92932 80568 29048 48750 42698 87434 83270 91766 84342 61161 3281 7470 
OK

5
63411 58718 5485 52733 14271 
OK

4
62921 71105 25136 46591 
OK

12
74981 99821 59633 87319 

# Screenhsoots of the Coursera Submissions in C++ and Python

Sum of two digits C++ and python


![AplusB](https://raw.githubusercontent.com/gimoralesg/AlgorithmsUN2021II/main/Lab7/pantallazo_ej1courseraAT.png)

![AplusB](https://raw.githubusercontent.com/gimoralesg/AlgorithmsUN2021II/main/Lab7/pantallazo_sumA_B.png)

Maximum pairwise product error
![maxproderror](https://raw.githubusercontent.com/gimoralesg/AlgorithmsUN2021II/main/Lab7/max_product_error.png)

Maximum pairwise product C++ and Python

![maxprodc++](https://raw.githubusercontent.com/gimoralesg/AlgorithmsUN2021II/main/Lab7/max_product_passed_cpp.png)

![maxprodpy](https://raw.githubusercontent.com/gimoralesg/AlgorithmsUN2021II/main/Lab7/max_product_passed_py.png)